### ISSUE TITLE
std::merkle functions break for path array of length >254

### ISSUE THREAD

**Issue by Philogy** [noir-lang/noir#1129]
### Aim

I tried to implement a sparse merkle tree using `std::merkle::compute_root_from_leaf` and `std::merkle::check_membership`. However, the compiler gives a strange error when I try to use `std::merkle::check_membership`:

**`main.nr`:**
```noir
mod smt;

fn main(start_root: pub Field, value: pub Field, path: [Field; smt::SMT_SIZE]) -> pub Field {
  smt::insert(start_root,_ value, path)
}
``` 

**`smt.nr`:**
```noir
use dep::std;

global SMT_SIZE: Field = 256;

fn insert(start_root: Field, value: Field, path: [Field; SMT_SIZE]) -> Field {
  constrain std::merkle::check_membership(start_root, 0, value, path) == 1;
  std::merkle::compute_root_from_leaf(1, value, path)
}

fn delete(start_root: Field, value: Field, path: [Field; SMT_SIZE]) -> Field {
  constrain std::merkle::compute_root_from_leaf(1, value, path) == start_root;
  std::merkle::compute_root_from_leaf(0, value, path)
}
```

Substituting `check_membership` for `compute_root_from_leaf` also does not work but results in a different error and works with `SMT_SIZE` up to 253 but breaks when trying to get the full size of 256

### Expected behavior

I expected the code to compile with no errors.

### Bug

For `check_membership` I get the following error when `SMT_SIZE` is `253` or below:

```
The application panicked (crashed).
Message:  internal error: entered unreachable code: invalid input: Const(Constant { id: NodeId(Index { index: 775, generation: 0 }), value: 0, value_str: ""
, value_type: NativeField })
Location: crates/noirc_evaluator/src/ssa/acir_gen/operations/intrinsics.rs:188

This is a bug. Consider opening an issue at https://github.com/noir-lang/noir/issues/new?labels=bug&template=bug_report.yml
```

For both variants `compute_root_from_leaf(...) == root` and `check_membership(...) == 1` the compiler breaks when the `SMT_SIZE` is set to above 253. 

### To reproduce

1. `nargo new your-project`
2. Create `main.nr`:
```noir
mod smt;

fn main(start_root: pub Field, value: pub Field, path: [Field; smt::SMT_SIZE]) -> pub Field {
  smt::insert(start_root, value, path)
}
```
3. Create `smt.nr`:
```noir
use dep::std;

global SMT_SIZE: Field = 256;

fn insert(start_root: Field, value: Field, path: [Field; SMT_SIZE]) -> Field {
  constrain std::merkle::compute_root_from_leaf(0, value, path) == start_root;
  std::merkle::compute_root_from_leaf(1, value, path)
}

fn delete(start_root: Field, value: Field, path: [Field; SMT_SIZE]) -> Field {
  constrain std::merkle::compute_root_from_leaf(1, value, path) == start_root;
  std::merkle::compute_root_from_leaf(0, value, path)
}
```
4. Run `nargo compile`

### Installation method

Binary

### Nargo version

nargo 0.3.2 (git version hash: 29b1f7df4d563849a62e64c533cb62932188135b, is dirty: false)

### Submission Checklist

- [ ] Once I hit submit, I will assign this issue to the Project Board with the appropriate tags.

**Comment by Philogy** [comment]
Since the field size is 254 bits large my sparse merkle tree would need a proof path size of at least 254 :/

**Comment by Philogy** [comment]
@TomAFrench thx for explaining.

Is this something that can be fixed i.e. allowing `Field.to_le_bits` to decompose to 254 bits or is this a limitation we'll have to live with?

Couldn't `to_le_bits` just additionally constrain that the value is being decomposed into the lower (non-overflowing) representation if `size >= 254` since that parameter has to be known at compile-time anyway?

Thinking about it depending on the situation it may be safe to just use 253 bits rather than field's full bit size of 254 if the user can't directly choose the `index` value as it's computationally infeasible to find a collision in 126.5 bits (halved bc of birthday attack) for now.

**Comment by TomAFrench** [comment]
> Is this something that can be fixed i.e. allowing Field.to_le_bits to decompose to 254 bits or is this a limitation we'll have to live with?
>
> Couldn't `to_le_bits` just additionally constrain that the value is being decomposed into the lower (non-overflowing) representation if size >= 254 since that parameter has to be known at compile-time anyway?

Yes, we'd need to think about the most efficient way of doing it however. In the worst case scenario, you can always decompose the index into a bit array inside noir itself however this is quite inefficient which is why we have this optimized version.

Using 253 bits for now seems fine imo while you're experimenting but we'll want to address this issue (along with the errors when using the `check_membership` function.)

Flagging this issue up to @joss-aztec as he's working on a SMT implementations as well so may be able to provide some insight.

**Comment by TomAFrench** [comment]
Hey @Philogy, just wanted to let you know that this has been addressed in #1312. I've tested your code again and got it compiling.*

* It seems like there's a bit of an issue with passing constant values into `compute_merkle_root` so your code doesn't compile exactly as it is due to the hardcoded leaf values but that shouldn't be an issue in real usage.

I've opened an issue for it here #1345 and am going to close this issue as resolved.
